cache/fast_lru_cache.cc

```diff
@@ -369,26 +351,43 @@ Status LRUCacheShard::Insert(const Slice& key, uint32_t hash, void* value,
     if ((usage_ + tmp.total_charge > capacity_ &&
          (strict_capacity_limit_ || handle == nullptr)) ||
         table_.GetOccupancy() == size_t{1} << table_.GetLengthBits()) {
-      // Originally, when strict_capacity_limit_ == false and handle != nullptr
-      // (i.e., the user wants to immediately get a reference to the new
-      // handle), the insertion would proceed even if the total charge already
-      // exceeds capacity. We can't do this now, because we can't physically
-      // insert a new handle when the table is at maximum occupancy.
+      // There are two measures of capacity:
+      // - Space (or charge) capacity: The maximum possible sum of the charges
+      //    of the elements.
+      // - Table capacity: The number of slots in the hash table.
+      // These are incomparable, in the sense that one doesn't imply the other.
+      // Typically we will reach space capacity before table capacity---
+      // if the user always inserts values with size equal to
+      // estimated_value_size, then at most a kLoadFactor fraction of slots
+      // will ever be occupied. But in some cases we may reach table capacity
+      // before space capacity---if the user initially claims a very large
+      // estimated_value_size but then inserts tiny values, more elements than
+      // initially estimated will be inserted.
+
       // TODO(Guido) Some tests (at least two from cache_test, as well as the
-      // stress tests) currently assume the old behavior.
+      // stress tests) currently assume the table capacity is unbounded.
       if (handle == nullptr) {
         // Don't insert the entry but still return ok, as if the entry inserted
         // into cache and get evicted immediately.
         last_reference_list.push_back(tmp);
       } else {
-        s = Status::Incomplete("Insert failed due to LRU cache being full.");
+        if (table_.GetOccupancy() == size_t{1} << table_.GetLengthBits()) {
```

Comment:
I worry about performance impact of allowing table to fill up. Seems better to me to enforce some less-than-full limit and return failure if that is reached.